kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+
+- **Easier overview of user balance** - because all ETH is in one place we can easily give the user an on-chain overview of their total deposits and payouts on Kickback.
+
+- **Upgrade-able contracts** - by making the user pot upgradeable (with admin approval) we allow for flexibility in future and for bug fixing capability.
+
+- **Configurable event incentives** - one of the nice things about this architecture is that the `UserPot` is only responsible for summing up deposits and payouts - actual calculation of what the payout should be is done by the `Event` contract. If we wished to test out new strategies (e.g. sweepstakes) or try using a single `Event` instance for multiple events we could easily do so.
+
+- **Potential revenue opportunity** - with the rise of tools and projects such as rDAI and the DAI savings rate there is increasing opportunity of earning interest from staked assets. By changing the structure to hold our customers money it will increase the overall revenue potential when we integrate with such features in future.
+
+## Risks and and trade-offs
+
+Although gas costs for RSVP'ing have gone up, I think this is justifiable given the improvements in overall user experience, plus the additional flexibility we gain over our current model.
+
+The events emitted by the event contract will all stay the same - meaning our backend doesn't need to change enormously for this. However we will no longer show when a user has withdrawn from an event - there's no need to since withdrawals across events are now batched.
+
+Storing all ETH into a single contract may seem risky, but we can negate this with good testing and auditing of the contract, and by managing admin access using multisig accounts with op-sec supervision.
+
+## Implementation
+
+_Note: the code below does not take into account ERC-20 funded events, just ETH ones_
+
+**EternalStorage**
+
+The base class which specifies an extendable storage system that allows for easy upgrades.
+
+```solidity
+/**
+ * Base storage class.
+ */
+contract EternalStorage {
+  // scalars
+  mapping(string => address) dataAddress;
+  mapping(string => string) dataString;
+  mapping(string => bytes32) dataBytes32;
+  mapping(string => int256) dataInt256;
+  mapping(string => uint256) dataUint256;
+  mapping(string => bool) dataBool;
+  // arrays
+  mapping(string => address[]) dataManyAddresses;
+  mapping(string => bytes32[]) dataManyBytes32s;
+  mapping(string => int256[]) dataManyInt256;
+  mapping(string => uint256[]) dataManyUint256;
+  mapping(string => bool[]) dataManyBool;
+}
+```
+
+**Proxy**
+
+Base class for all upgradeable contracts. Upgradeability is based on the [OpenZepellin Proxy + EternalStorage](https://github.com/OpenZeppelin/openzeppelin-labs/tree/master/upgradeability_using_eternal_storage) pattern.
```

Comment:
When actually implementing it, is it better to just use https://openzeppelin.com/sdk/ ?